nemo/collections/nlp/models/nlp_model.py

```diff
@@ -312,53 +252,225 @@ def setup(self, stage: str) -> None:
-                        )
-                        self.bert_model.restore_weights(self.bert_model._restore_path)
-
-                    logging.info("Replacing sampler with model parallel sampler")
-                    mp_sampler = torch.utils.data.distributed.DistributedSampler(
-                        self._train_dl.dataset,
-                        num_replicas=app_state.data_parallel_size,
-                        rank=app_state.data_parallel_rank,
-                    )
-                    mp_dl = self._trainer.replace_sampler(self._train_dl, mp_sampler)
-                    self._train_dl = mp_dl
-                else:
-                    raise NotImplementedError(
-                        f'The BERT encoder: {self.bert_model} does not support model parallelism yet.'
-                    )
-
     def on_save_checkpoint(self, checkpoint: Dict[str, Any]) -> None:
         if hasattr(self, "bert_model") and isinstance(self.bert_model, MegatronBertEncoder):
             checkpoint['checkpoint_version'] = get_checkpoint_version()
         return None
 
+    def on_load_checkpoint(self, checkpoint: Dict[str, Any]) -> None:
+        if hasattr(self, "bert_model") and isinstance(self.bert_model, MegatronBertEncoder):
+            set_checkpoint_version(checkpoint['checkpoint_version'])
+            logging.info(f"Setting Megatron checkpoint version: {checkpoint['checkpoint_version']}")
+        return None
+
+    # no rank check as model parallel models need to be saved on data parallel rank 0
+    def save_to(self, save_path: str):
+        """
+        Saves model instance (weights and configuration) into .nemo file
+         You can use "restore_from" method to fully restore instance from .nemo file.
+
+        .nemo file is an archive (tar.gz) with the following:
+            model_config.yaml - model configuration in .yaml format. You can deserialize this into cfg argument for model's constructor
+            model_wights.chpt - model checkpoint
+
+        Args:
+            save_path: Path to .nemo file where model instance should be saved
+        """
+
+        app_state = AppState()
+        if app_state.model_parallel_size is not None:
+            self._default_save_to(save_path)
+        else:
+            return super().save_to(save_path)
+
+    def _default_save_to(self, save_path: str):
+        app_state = AppState()
+        if app_state.model_parallel_size is not None:
+            # each model parallel rank creates a .nemo file
+            # after all .nemo files are created, each rank
+            # will add their checkpoint to global rank 0
+
+            base_dir = os.path.dirname(save_path)  # use the directory to merge mp_rank .nemo files into one
+
+            # TODO: accept general paths
+            # update save_path based on model parallel_rank
+            base_path = save_path[0:-5]  # everything excpe the .nemo extension
```

Comment:
Lets correct this before merging. os.path.splitext(X)[0] should work here ?

Reply:
Yes, I will fix this before we merge.

Reply:
Done.